src/os/bluestore/BlueStore.cc

```diff
@@ -6987,52 +7841,72 @@ int BlueStore::_clone(TransContext *txc,
     goto out;
 
   if (g_conf->bluestore_clone_cow) {
-    if (!oldo->onode.extent_map.empty()) {
-      if (!oldo->bnode) {
-	oldo->bnode = c->get_bnode(newo->oid.hobj.get_hash());
-      }
-      if (!newo->bnode) {
-	newo->bnode = oldo->bnode;
-      }
-      assert(newo->bnode == oldo->bnode);
-      // move blobs
-      map<int64_t,int64_t> moved_blobs;
-      for (auto& p : oldo->onode.extent_map) {
-        if (!p.second.is_shared()) {
-          BlobRef b;
-          if (moved_blobs.count(p.second.blob) == 0) {
-            b = oldo->blob_map.get(p.second.blob);
-            oldo->blob_map.erase(b);
-            newo->bnode->blob_map.claim(b);
-            moved_blobs[p.second.blob] = b->id;
-            dout(30) << __func__ << "  moving old onode blob " << p.second.blob
-                    << " to bnode blob " << b->id << dendl;
-            b->dirty_blob().clear_flag(bluestore_blob_t::FLAG_MUTABLE);
-            b->dirty_blob().set_flag(bluestore_blob_t::FLAG_HAS_REFMAP);
-          } else {
-            b = newo->bnode->blob_map.get(moved_blobs[p.second.blob]);
-          }
-          b->dirty_blob().get_ref(p.second.offset, p.second.length);
-        }
-      }
-      // update lextents
-      for (auto& p : oldo->onode.extent_map) {
-	if (moved_blobs.count(p.second.blob)) {
-	  p.second.blob = -moved_blobs[p.second.blob];
+    // hmm, this could go into an ExtentMap::dup() method.
+    vector<BlobRef> id_to_blob(oldo->extent_map.extent_map.size());
+    for (auto &e : oldo->extent_map.extent_map) {
+      e.blob->last_encoded_id = -1;
+    }
+    int n = 0;
+    bool dirtied_oldo = false;
+    for (auto &e : oldo->extent_map.extent_map) {
+      BlobRef nb;
+      if (e.blob->last_encoded_id >= 0) {
+	// blob is already duped
+	nb = id_to_blob[e.blob->last_encoded_id];
+      } else {
+	// dup the blob
+	const bluestore_blob_t& blob = e.blob->get_blob();
+	// make sure it is shared
+	if (!blob.is_shared()) {
+	  e.blob->dirty_blob().sbid = _assign_blobid(txc);
+	  c->make_blob_shared(e.blob);
+	  dirtied_oldo = true;  // fixme: overkill
+	} else if (!e.blob->shared_blob->loaded) {
+	  c->load_shared_blob(e.blob->shared_blob);
+	}
+	nb = new Blob;
```

Comment:
This reads dangerously, because nb is sometimes used for existing blobs and sometimes for a new blob. Perhaps a helper function that creates blobs and returns the reference to the created blob?